cocoa/src/toga_cocoa/factory.py

```diff
@@ -1,90 +1,55 @@
 
-from . import dialogs
-from .app import App, DocumentApp, MainWindow
-from .command import Command
-from .documents import Document
-from .fonts import Font
-from .hardware.camera import Camera
-from .hardware.location import Location
-from .icons import Icon
-from .images import Image
-from .paths import Paths
+from toga import NotImplementedWarning
 
-# Widgets
-from .widgets.activityindicator import ActivityIndicator
-from .widgets.box import Box
-from .widgets.button import Button
-from .widgets.canvas import Canvas
-from .widgets.detailedlist import DetailedList
-from .widgets.divider import Divider
-from .widgets.imageview import ImageView
-from .widgets.label import Label
-from .widgets.mapview import MapView
-from .widgets.multilinetextinput import MultilineTextInput
-from .widgets.numberinput import NumberInput
-from .widgets.optioncontainer import OptionContainer
-from .widgets.passwordinput import PasswordInput
-from .widgets.progressbar import ProgressBar
-from .widgets.scrollcontainer import ScrollContainer
-from .widgets.selection import Selection
-from .widgets.slider import Slider
-from .widgets.splitcontainer import SplitContainer
-from .widgets.switch import Switch
-from .widgets.table import Table
-from .widgets.textinput import TextInput
-from .widgets.tree import Tree
-from .widgets.webview import WebView
-from .window import Window
+toga_factory_imports = {
+    "Window": "toga.window",
+    "WebView": "toga.widgets.webview",
+    "Tree": "toga.widgets.tree",
+    "TextInput": "toga.widgets.textinput",
+    "Table": "toga.widgets.table",
+    "Switch": "toga.widgets.switch",
+    "SplitContainer": "toga.widgets.splitcontainer",
+    "Slider": "toga.widgets.slider",
+    "Selection": "toga.widgets.selection",
+    "ScrollContainer": "toga.widgets.scrollcontainer",
+    "ProgressBar": "toga.widgets.progressbar",
+    "PasswordInput": "toga.widgets.passwordinput",
+    "OptionContainer": "toga.widgets.optioncontainer",
+    "NumberInput": "toga.widgets.numberinput",
+    "MultilineTextInput": "toga.widgets.multilinetextinput",
+    "MapView": "toga.widgets.mapview",
+    "Label": "toga.widgets.label",
+    "ImageView": "toga.widgets.imageview",
+    "Divider": "toga.widgets.divider",
+    "DetailedList": "toga.widgets.detailedlist",
+    "Canvas": "toga.widgets.canvas",
+    "Button": "toga.widgets.button",
+    "Box": "toga.widgets.box",
+    "ActivityIndicator": "toga.widgets.activityindicator",
+    "Path": "toga.paths",
+    "Image": "toga.images",
+    "Icon": "toga.icons",
+    "Location": "toga.hardware.location",
+    "Camera": "toga.hardware.camera",
+    "Font": "toga.fonts",
+    "Document": "toga.documents",
+    "Command": "toga.command",
+    "dialogs": "toga_cocoa",
+    "App": "toga.app",
+    "DocumentApp": "toga.app",
+    "DocumentMainWindow": "toga.app",
+}
+
+
+def __getattr__(name):
+    if name in toga_factory_imports:
+        module = importlib.import_module(f"{toga_factory_imports[name]}")
```

Comment:
It would be better to avoid repeated lookups by using local variables:

```py
try:
    module_name = toga_factory_imports[name]
except KeyError:
    raise NotImplementedError(f"Toga's Cocoa backend doesn't implement {name}") from None
else:
    module = importlib.import_module(module_name)
    value = getattr(module, name)
    globals()[name] = value
    return value
```